src/compiler/crystal/zero_one_or_many.cr

```diff
@@ -0,0 +1,110 @@
+# An Array(T)-like type that's optimized for the case of
+# frequently having zero or one elements.
+#
+# It's a struct, so if you have a ZeroOneOrMany and you
+# want to mutate, you have to use `with` or `without`
+# and overwrite the value you have with what you get
+# from those methods.
+#
+# Warning: you should never keep the original value
+# around when using the above methods as the internal
+# array might be mutated (for performance reasons),
+# so the original value might mutate in that case!
+#
+# TODO: add tests for this type.
+struct Crystal::ZeroOneOrMany(T)
+  include Enumerable(T)
+
+  getter value : Nil | T | Array(T)
+
+  def initialize
+    @value = nil
+  end
+
+  def initialize(@value : T)
+  end
+
+  def initialize(values : Array(T))
+    @value =
+      case values.size
+      when 0
+        nil
+      when 1
+        values.first
+      else
+        values
+      end
+  end
+
+  def each(& : T ->)
+    value = @value
+    case value
+    in Nil
+      # Nothing to do
+    in T
+      yield value
+    in Array(T)
+      value.each do |element|
+        yield element
+      end
+    end
+  end
+
+  def size : Int32
+    value = @value
+    case value
+    in Nil
+      0
+    in T
+      1
+    in Array(T)
+      value.size
+    end
+  end
+
+  def with(element : T) : ZeroOneOrMany(T)
+    value = @value
+    case value
+    in Nil
+      ZeroOneOrMany(T).new(element)
+    in T
+      ZeroOneOrMany(T).new([value, element] of T)
+    in Array(T)
+      value.push element
+      self
+    end
+  end
+
+  def with(elements : Enumerable(T)) : ZeroOneOrMany(T)
+    value = @value
+    case value
+    in Nil
+      ZeroOneOrMany(T).new(elements.map(&.as(T)))
+    in T
+      new_elements = Array(T).new(elements.size + 1)
+      new_elements.push(value)
+      new_elements.concat(elements)
+      ZeroOneOrMany.new(new_elements)
```

Comment:
Could it be an option to reuse `elements` if it's an `Array(T)` (or `ZeroOneOrMany(T)` with an array value) instead of allocating a new array all the time? This would apply the same constraint for forgetting about the original data at the callsite. Might squeeze out some more performance.

Reply:
I wanted to do that, but I couldn't find a way to express that with the type system and restrictions we have. Can you think of a way to do it? Feel free to experiment with this branch. We could also leave it as an optimization to do later on.

Reply:
Also in the end it might not be good to reuse the array as it might later be mutated, affecting the original value.

Reply:
Hm, what were you having difficulties with?
I was thinking of something like this:
```cr
if elements.is_a?(Array(T))
  elements.unshift value
  ZeroOneOrMany.new(elements)
else
  new_elements = Array(T).new(elements.size + 1)
  new_elements.push(value)
  new_elements.concat(elements)
  ZeroOneOrMany.new(new_elements)
end
```
Although, I suppose the `unshift` would quite likely re-allocate anyways because incoming arrays are probably always aligned to the start.

Reply:
Ah, that might work.

That said, the array will then be retained and if the ZeroOneOrMany is further used and the array is mutated, it will be visible to the outside world. So maybe not doing this will be fine.

I'm also now more convinced that making the type explicitly mutable is better. See https://github.com/crystal-lang/crystal/pull/12426